aten/src/ATen/cpu/vec/vec512/vec512.h

```diff
@@ -274,6 +274,17 @@ inline Vectorized<uint8_t> flip(const Vectorized<uint8_t> & v) {
   return flip8(v);
 }
 
+inline Vectorized<bool> operator&&(
+    const Vectorized<bool>& self,
+    const Vectorized<bool>& other) {
+  const __m512i* self_ = reinterpret_cast<const __m512i*>(self.as_bytes());
+  const __m512i* other_ = reinterpret_cast<const __m512i*>(other.as_bytes());
+  __m512i out = _mm512_and_si512(*self_, *other_);
+  Vectorized<bool> ret;
+  std::memcpy(ret, &out, ret.size() * sizeof(bool));
```

Comment:
A bit confused why you want to do memory copies here. What's the problem if we do `_mm512_and_si512` directly on two vectors? Would your implementation be ever faster than the scalar version?

Reply:
`_mm512_and_si512` happens on two `_mm512i`and return a `__mm512i`.
We can use `reinterpret_cast` to convert `Vectorized<bool>` to `__mm512i`.
 
Currently there is no constructor from `_mm512i` to `Vectorized<bool>`, I take the reference here https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cpu/vec/vec.h#L19 to convert return a `Vectorized<bool>`.

As for performance, I observed 4x for this shape `x = torch.ones((10240), dtype=torch.bool)` on ICX. ( Larger shape to avoid overhead become a major factor)

Reply:
I'm curious why no other function here needed this reinterpret_cast/memcpy but this one does?

Reply:
> `_mm512_and_si512` happens on two `_mm512i`and return a `__mm512i`. We can use `reinterpret_cast` to convert `Vectorized<bool>` to `__mm512i`.
> 
> Currently there is no constructor from `_mm512i` to `Vectorized<bool>`, I take the reference here https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cpu/vec/vec.h#L19 to convert return a `Vectorized<bool>`.
> 
> As for performance, I observed 4x for this shape `x = torch.ones((10240), dtype=torch.bool)` on ICX. ( Larger shape to avoid overhead become a major factor)

OK. The code you referred to try to make sure the boolean values are valid but here the result should already be valid given the inputs are valid booleans. In fact, we don't have intrinsic-based implementation for boolean vectors. Supporting `__mm512i` as constructor seems requiring more code changes. I'm open to do memcpy here or adding boolean intrinsic classes. If you choose the former as what you are doing here, perhaps you can add a comment to explain why you are doing memcpy here.

Reply:
> I'm curious why no other function here needed this reinterpret_cast/memcpy but this one does?

Hi, @albanD.
For "other functions", do you mean some functions here
https://github.com/pytorch/pytorch/blob/f7eae279463b719c5f25587aac225bd2be891373/aten/src/ATen/cpu/vec/vec512/vec512_int.h#L292-L294
This functions takes a `Vectorized<int32_t>` and return a `Vectorized<int32_t>` without `reinterpret_cast/memcpy` because constructer here https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cpu/vec/vec512/vec512_int.h#L27 and here https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cpu/vec/vec512/vec512_int.h#L19.
We can also avoid `reinterpret_cast/memcpy` for adding boolean intrinsic classes `Vectorized<bool>`. Following this but will introduce more code changes, and since we actually don't have intrinsic-based implementation for boolean vectors, and we also have speed up here for vec impl. We may not need to adding it.

Do you have any prefers on it?

Reply:
If we expect these vectorized implementations might be needed in other places, we should add them yes.
Otherwise, what we have here is ok.

Reply:
For now, I cannot see other cases need a `Vectorized<bool>`.
And from here https://github.com/pytorch/pytorch/blob/8a2fed7e6ab4cb6a97d92f6ca14f257370ec3a92/aten/src/ATen/cpu/vec/vec_base.h#L127
Adding `Vectorized<bool>` should be a lot code changes and in my opinion we can leave this PR as it is and implement `Vectorized<bool>` after we found more cases need to do so.

Reply:
> > `_mm512_and_si512` happens on two `_mm512i`and return a `__mm512i`. We can use `reinterpret_cast` to convert `Vectorized<bool>` to `__mm512i`.
> > Currently there is no constructor from `_mm512i` to `Vectorized<bool>`, I take the reference here https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cpu/vec/vec.h#L19 to convert return a `Vectorized<bool>`.
> > As for performance, I observed 4x for this shape `x = torch.ones((10240), dtype=torch.bool)` on ICX. ( Larger shape to avoid overhead become a major factor)
> 
> OK. The code you referred to try to make sure the boolean values are valid but here the result should already be valid given the inputs are valid booleans. In fact, we don't have intrinsic-based implementation for boolean vectors. Supporting `__mm512i` as constructor seems requiring more code changes. I'm open to do memcpy here or adding boolean intrinsic classes. If you choose the former as what you are doing here, perhaps you can add a comment to explain why you are doing memcpy here.

Thanks for advice, comments added.